File: engine/src/main/java/com/terralite/engine/terrain/CompactBlockStorage.java
package com.terralite.engine.terrain;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.IntFunction;
import java.util.function.ToIntFunction;

public final class CompactBlockStorage implements BlockStorage {
    private final Map<BlockPos, Integer> blocks = new LinkedHashMap<>();
    private final ToIntFunction<BlockState> stateIds;
    private final IntFunction<BlockState> states;
    private final int airStateId;

    public CompactBlockStorage(
            ToIntFunction<BlockState> stateIds,
            IntFunction<BlockState> states,
            int airStateId
    ) {
        this.stateIds = Objects.requireNonNull(stateIds, "stateIds");
        this.states = Objects.requireNonNull(states, "states");
        this.airStateId = airStateId;
    }

    @Override
    public BlockState get(BlockPos pos) {
        Integer stateId = blocks.get(Objects.requireNonNull(pos, "pos"));
        return stateId == null ? BlockState.AIR : states.apply(stateId);
    }

    public int getStateId(BlockPos pos) {
        return blocks.getOrDefault(Objects.requireNonNull(pos, "pos"), airStateId);
    }

    @Override
    public void set(BlockPos pos, BlockState state) {
        Objects.requireNonNull(pos, "pos");
        Objects.requireNonNull(state, "state");
        int stateId = stateIds.applyAsInt(state);
        setStateId(pos, stateId);
    }

    public void setStateId(BlockPos pos, int stateId) {
        Objects.requireNonNull(pos, "pos");
        if (stateId == airStateId || states.apply(stateId).isAir()) {
            blocks.remove(pos);
            return;
        }
        blocks.put(pos, stateId);
    }

    @Override
    public BlockState remove(BlockPos pos) {
        Integer removed = blocks.remove(Objects.requireNonNull(pos, "pos"));
        return removed == null ? BlockState.AIR : states.apply(removed);
    }

    public int removeStateId(BlockPos pos) {
        Integer removed = blocks.remove(Objects.requireNonNull(pos, "pos"));
        return removed == null ? airStateId : removed;
    }

    @Override
    public boolean contains(BlockPos pos) {
        return blocks.containsKey(Objects.requireNonNull(pos, "pos"));
    }

    @Override
    public Collection<BlockPos> positions() {
        return List.copyOf(blocks.keySet());
    }

    @Override
    public int size() {
        return blocks.size();
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/terrain/CompactBlockStorage.kt
package com.terralite.engine.terrain

class CompactBlockStorage(
    private val stateIds: (BlockState) -> Int,
    private val states: (Int) -> BlockState,
    private val airStateId: Int
) : BlockStorage {
    private val blocks: MutableMap<BlockPos, Int> = LinkedHashMap()

    override fun get(pos: BlockPos): BlockState {
        val stateId = blocks[pos] ?: return BlockState.AIR
        return states(stateId)
    }

    fun getStateId(pos: BlockPos): Int = blocks.getOrDefault(pos, airStateId)

    override fun stateId(pos: BlockPos): Int = getStateId(pos)

    override fun set(pos: BlockPos, state: BlockState) {
        setStateId(pos, stateIds(state))
    }

    fun setStateId(pos: BlockPos, stateId: Int) {
        if (stateId == airStateId || states(stateId).isAir()) {
            blocks.remove(pos)
        } else {
            blocks[pos] = stateId
        }
    }

    override fun remove(pos: BlockPos): BlockState {
        val removed = blocks.remove(pos) ?: return BlockState.AIR
        return states(removed)
    }

    fun removeStateId(pos: BlockPos): Int = blocks.remove(pos) ?: airStateId

    override fun contains(pos: BlockPos): Boolean = blocks.containsKey(pos)

    override fun positions(): Collection<BlockPos> = blocks.keys.toList()

    override fun size(): Int = blocks.size
}
